### ISSUE TITLE
Allow defining callbacks to execute on transaction commit

### ISSUE THREAD

**Issue by sds** [kysely-org/kysely#396]
**TL;DR**: Add a `onCommit` method to [`Transaction`](https://github.com/kysely-org/kysely/blob/8a489c2cc952885543088fdf33e3e900d35315e5/src/kysely.ts#L346) which can be passed multiple callbacks to execute if and only if the transaction commits.

If you're open to the feature, I'm happy to put up a PR with the proposed solution.

### Motivation

When passing a `Transaction` to helper functions, it can be useful define logic which executes only if the transaction commits successfully. For example, specifying side effects like sending emails or push notifications where such side effects should not be carried out if the transaction is rolled back.

Normally, you would just do this after the transaction:
```ts
await db.transaction().execute(async (trx) => {
  ...
});

// If we make it here, transaction was successful, so carry out side effects
sendEmail(...);
```

However, in situations where you want to offer reusable helpers that perform logic within the context of a transaction, allowing those helpers to define callbacks to execute on successful transaction commit can help DRY up code significantly, and make it easier to ensure desired side effects are consistently carried out.

#### Example

```ts
// Example of using multiple helpers taking the same Transaction object
await db.transaction().execute(async (trx) => {
  await updateBalance(trx, account, balance);
  await logAuditEvent(trx, { account, severity: 'info', action: `Account refunded` });
});

/////////////////////////////////////////////////////////////////////////////////////////////////
// Helpers reused in multiple places
/////////////////////////////////////////////////////////////////////////////////////////////////

export const logAuditEvent = async (trx: Transaction<Database>, event: AuditEvent) => {
  const record = trx
    .insertInto('audit_events')
    .values(event)
    .returningAll()
    .executeTakeFirstOrThrow();

  if (record.severity === 'alert') {
    trx.onCommit(() => { sendAlertEmail(record.account); });
  }
};

export const updateBalance = async (trx: Transaction<Database>, account: string, balance: number) => {
  const record = trx
    .updateTable('balances')
    .where('account', '=', account)
    .set({ balance })
    .returningAll()
    .executeTakeFirstOrThrow();

  if (record.balance < 10) {
    trx.onCommit(() => { sendLowBalanceEmail(account, record.balance); });
  }
};
```

Benefits of this pattern are:
- Ability to define reusable components with side effects.
- Access to values available in the closure in which it is defined (useful in deeply nested helpers).
- Easier than manually passing around an `EventEmitter` to manually aggregate.

### Proposal

At a high level, one potential implementation approach would be:

- Add an `onCommit` method to `Transaction`.
- Update `Transaction` to create an [`EventEmitter`](https://nodejs.org/api/events.html#class-eventemitter) only if `onCommit` is called (so zero memory overhead if this feature isn't used) and use [`once`](https://nodejs.org/api/events.html#emitteronceeventname-listener) to register any callbacks each time `onCommit` is called. 
- Add an `override` method to `Transaction` for [`execute`](https://github.com/kysely-org/kysely/blob/8a489c2cc952885543088fdf33e3e900d35315e5/src/kysely.ts#L458-L471) which calls `emit` after executing the underlying transaction successfully.

### Workaround

If this functionality isn't the right fit for Kysely, anyone interested in implementing a similar pattern can do something like the following:

```ts
import { Kysely, Transaction } from 'kysely';
import { TypedEmitter } from 'tiny-typed-emitter';

interface TransactionEvents {
  commit: () => void;
}

class TransactionEventEmitter extends TypedEmitter<TransactionEvents> {
  constructor() {
    super();
  }
}

declare module 'kysely/dist/cjs/kysely' {
  interface Kysely<DB> {
    executeTransaction<T>(fn: (trx: Transaction<DB>, hooks: TransactionEventEmitter) => Promise<T>): Promise<T>;
  }
}

Kysely.prototype.executeTransaction = async function (
  this: Kysely<any>,
  fn: (trx: Transaction<any>, hooks: TransactionEventEmitter) => Promise<any>
) {
  if (this.isTransaction) throw new Error('Cannot call executeTransaction on a transaction.');

  const hooks = new TransactionEventEmitter();

  const result = await this.transaction()
    .setIsolationLevel('serializable')
    .execute(async (trx) => {
      return fn(trx, hooks);
    });

  hooks.emit('commit');

  return result;
};
```

Using our earlier example:

```ts
// Example of using multiple helpers taking the same Transaction object
await db.executeTransaction(async (trx, hooks) => {
  await updateBalance(trx, hooks, account, balance);
  await logAuditEvent(trx, hooks, { account, severity: 'info', action: `Account refunded` });
});

/////////////////////////////////////////////////////////////////////////////////////////////////
// Helpers reused in multiple places
/////////////////////////////////////////////////////////////////////////////////////////////////

export const logAuditEvent = async (trx: Transaction<Database>, hooks: TransactionEventEmitter, event: AuditEvent) => {
  const record = trx
    .insertInto('audit_events')
    .values(event)
    .returningAll()
    .executeTakeFirstOrThrow();

  if (record.severity === 'alert') {
    hooks.on('commit', () => { sendAlertEmail(record.account); });
  }
};

export const updateBalance = async (trx: Transaction<Database>, hooks: TransactionEventEmitter, account: string, balance: number) => {
  const record = trx
    .updateTable('balances')
    .where('account', '=', account)
    .set({ balance })
    .returningAll()
    .executeTakeFirstOrThrow();

  if (record.balance < 10) {
    hooks.on('commit', () => { sendLowBalanceEmail(account, record.balance); });
  }
};
```

Unfortunately this gets a little verbose since we need to pass both the transaction and the event emitter everywhere. This is why it would be really nice if `onCommit` could be called on `Transaction` itself.

Even if we decide not to include this functionality in Kysely, hope others find the suggested workaround useful.

**Comment by koskimas** [comment]
Couldn't you just extend `Transaction` yourself? Everything is immutable in Kysely, so we could consider adding something like `trx = trx.onCommit(callback)` but I feel like it's not something a query builder should have.

**Comment by koskimas** [comment]
Actually I'm sure that's not a query builder's job. You can extend `Transaction` to have that method yourself.

**Comment by thhermansen** [comment]
> You can create a subclass of Transaction that has the method.

We are looking in to similar use case where it sometimes is practical to attach some function to be called after commit. 
@koskimas: How would you go around implementing this in a clean way?

I have taken a quick look in the src and found that extending `Kysely` to override `transaction()` isn't straight forward, as `TransactionBuilder` is to be passed the private `props` instance variable. And the 

https://github.com/kysely-org/kysely/blob/3d5e0e59d5536e2dbfee8e39ebecd0b5d981536e/src/kysely.ts#L253-L255

I think if it had been protected it would have been easier to extend. Or is there another way I'm not thinking of? Maybe drop down a to driver somehow which the TransactionBuilder calls within `execute` in order to start and commit transactions.

**Comment by bombillazo** [comment]
Yeah, extending the Kysely Client sometimes requires extending internal classes that ate private props of the Kysely instance. For example, we wanted to extend the Kysely class with some transaction helpers but could not because we needed to rewrite the whole class since we couldn't extend or access private properties from the child class.

**Comment by neil176** [comment]
@koskimas  Is there any chance of revisiting this or at least considering changing `#props` to no longer have the `#`? I think the recommended solution of extending the relevant classes is reasonable, but it's not really doable due to that private modifier.
